File: Data Structures & Algorithms/reverse-a-linked-list/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        ListNode* ptr = head;
        ListNode* prev = nullptr, *curr = nullptr;
        while(ptr && ptr->next) {
            curr = ptr;
            ptr = ptr->next;
            curr->next = prev;
            prev = curr;
        }
        if(ptr) ptr->next = curr;
        return ptr;
    }
};

File: Data Structures & Algorithms/reverse-nodes-in-k-group/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverse(self, prev, head, tail):
        if head.next == tail:
            prev.next = head
            return head

        newHead = self.reverse(prev, head.next, tail)
        head.next.next = head
        head.next = tail
        return newHead

    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        dummy = ListNode(0)
        dummy.next = head

        prev, slow, fast = dummy, head, head
        
        while fast:
            ctr = 0
            while fast and ctr < k:
                fast = fast.next
                ctr += 1

            if ctr < k:
                break

            self.reverse(prev, slow, fast)

            prev = slow
            slow = fast

        return dummy.next